Fix issue 3527 while importing some Twistlock scans

macedogm opened the pull request:
Fix issue #3527.

The following improvements were made:
1. Added two global variables to control the column displacement when importing the different type of Twistlock image scan.
2. Wrap finding's `title` and `component_name`, otherwise when they are too long the finding will fail to be imported in the database.
3. Properly validate if the finding severity is `moderate` or `Moderate`.
4. Cast `moderate` severity to `Medium` and `important` to `High`.

Please let me know if you agree with the proposed solution. Best regards.

valentijnscholten commented:
Thanks! Are you able to provide a sample scan of that other format as part of the unit test in `test_twistlock_parser.py`?

valentijnscholten commented:
and can you base it against `dev` please.

macedogm commented:
Yes, I will rebase it against `dev` and provide a sample scan. Will do it today.

macedogm commented:
@valentijnscholten Can you please have a look now?
I've added the unit tests. The integration tests are still failing, but I think that it's due to another reason than this PR.

valentijnscholten commented:
can you rebase or merge `dev` into the PR?

macedogm commented:
Rebased and all checks passed 👍🏻 
Thanks for the help @valentijnscholten !

madchap reviewed (approved):
lgtm, thanks :+1:

damiencarol commented:
@macedogm @madchap @valentijnscholten could it be better to use `csv` module? Like do this:
```
records = csv.DictReader(csv_file)
```
Because the files have headers. Could be better to rely on header column name

macedogm commented:
@damiencarol Good idea, makes total sense. I'll submit a PR using `csv.DictReader`. It should make the code smaller and works for both types of reports.

macedogm commented:
Following @damiencarol suggestion, I modified the Twistlock CSV parser to use use `csv.DictReader()`, so the same parser can work without modifications for both types of Twistlock reports.

One important thing to mention is that I removed the [`escape()`](https://github.com/DefectDojo/django-DefectDojo/pull/3532/files#diff-3176294e4014378ed7fcff6167cc0a0e5e077c2dd816ebed034e2c5c23e7e754L19) of the contents, since the the escaping doesn't seem to be done in all report parsers. Wouldn't be the case of moving it to the parts of the DefectDojo responsible for saving the findings in the database and handling the views? (Referencing [this](https://github.com/DefectDojo/django-DefectDojo/pull/2626#pullrequestreview-442591229) comment made by @valentijnscholten on another PR.

Can you all please review the new improvement and see if it makes sense? And if we must add an `escape()` on each field?

Best regards.

valentijnscholten commented on dojo/tools/twistlock/parser.py:
```diff
@@ -3,55 +3,39 @@
-        data_severity = self.get_field_from_row_or_default(row, severity_column, 'Info').capitalize()
-        data_cvss = self.get_field_from_row_or_default(row, cvss_column, '')
-        data_description = self.get_field_from_row_or_default(row, description_column, '')
+        data_vulnerability_id = row['CVE ID']
+        data_package_version = row['Package Version']
+        data_fix_status = row['Fix Status']
+        data_package_name = row['Packages']
+        data_id = row['Id']
+        data_severity = row['Severity'].capitalize()
+        data_cvss = row['CVSS']
+        data_description = description_column = row['Description']
```
what happens if any of these fields are not present. does the dictreader return some kind of default, or is a check needed?

damiencarol replied:
@valentijnscholten `csv.DictReader(csv_file)` return dictionnaries for items. So we could use `row.get('column', default_value)` if we want something more stable if one column is missing.

macedogm replied:
Validation for missing columns improved as suggested - `row.get('column', default_value)`.

valentijnscholten commented on dojo/api_v2/serializers.py:
```diff
@@ -1055,6 +1055,10 @@ def save(self, push_to_jira=False):
                 sev = item.severity
                 if sev == 'Information' or sev == 'Informational':
                     sev = 'Info'
+                elif sev.lower() == 'moderate':
+                    sev = 'Medium'
+                elif sev.lower() == 'important':
+                    sev = 'High'
```
might be best to handle all severities in the same way, i.e. convert to lowercase before comparison?

damiencarol replied:
1) I agree with @valentijnscholten comment. If you really want to handle `moderate` and `important` values in API v2 code, you should follow the same logic that handles `Information`/`Informational` values.
2) I think this change is not good. it will impact every parser. We should rely on the parsers to set the severity to the good values of the `dojo/model.py` file:
```Python
    SEVERITIES = {'Info': 4, 'Low': 3, 'Medium': 2,
                  'High': 1, 'Critical': 0}
```
Each tool have his own logic, the contract is that the parser set the value to one of `Info`, `Low`, `Medium`, `High` or `Critical`.

macedogm replied:
Created a [function](https://github.com/DefectDojo/django-DefectDojo/pull/3532/files#diff-3176294e4014378ed7fcff6167cc0a0e5e077c2dd816ebed034e2c5c23e7e754R148) to properly parse the severity. Modified both CSV and JSON parsers to use the function.

valentijnscholten commented on dojo/tools/twistlock/parser.py:
```diff
@@ -132,7 +112,7 @@ def get_item(vulnerability, test):
+        elif vulnerability['severity'].lower() == 'moderate':
             severity = "Medium"
         else:
             severity = vulnerability['severity'].title()
```
not sure what values title can have but the parser needs to make sure only to insert severities known to dojo, otherwise there can be unpredictable results.

damiencarol replied:
`.title()` do an uppercase for the first letter and a lower case for the rest of the string. It is a good function for severity in parsers code as for now, we have `Info`, `High`, etc... as values.
My advice here is just to encapsulate the logic of converting severity values from the format to Dojo values in a separate function. And just call it when creating the finding, like this:
```Python
severity = self.convert_severity(vulnerability['severity'])
```

macedogm replied:
Created a [function](https://github.com/DefectDojo/django-DefectDojo/pull/3532/files#diff-3176294e4014378ed7fcff6167cc0a0e5e077c2dd816ebed034e2c5c23e7e754R148) to properly parse the severity. Modified both CSV and JSON parser to use the function.

valentijnscholten commented:
Thanks, using some standard components is always good. I don't think we want/need to escape in the parser. Defect Dojo should not trust (but escape) any user provided data  in the database anyway upon rendering pages.

damiencarol commented on dojo/tools/twistlock/parser.py:
```diff
@@ -3,55 +3,39 @@
+        data_fix_status = row['Fix Status']
+        data_package_name = row['Packages']
+        data_id = row['Id']
+        data_severity = row['Severity'].capitalize()
```
`capitalize` is useless IMHO. See my comment on creating a function to manage severity values below.

macedogm replied:
Fixed in the conversation above.

damiencarol reviewed (changes requested):
The pull request is good and adds a lot of new unit tests.
The switch to use `textwrap` and `csv` module are good for me. I made some comments to improve the quality of the parsing logic for severities.
Tagging as `Request changes` as I think the modifications in `dojo/api_v2/serializers.py` are not aligned with the rest of the code.

macedogm commented:
Thanks for all the feedback. I'll work on the changes today and update the PR.

macedogm commented:
Reverted the modifications in `dojo/api_v2/serializers.py` in favour of using the function inside the parser to properly parse the severity.
Also added a validation for the [finding's title](https://github.com/DefectDojo/django-DefectDojo/pull/3532/files#diff-3176294e4014378ed7fcff6167cc0a0e5e077c2dd816ebed034e2c5c23e7e754R27-R32), since in some cases, mostly when the finding is related to a compliance check from Twistlock, it has no CVE and no package, so it's displayed as `:-` (image 1). If this case happens again, it will now display the description of the finding instead (image 2).

Image 1
![image](https://user-images.githubusercontent.com/17654553/103212559-67361e80-490b-11eb-83b5-c9f320bcc92f.png)

Image 2
![image](https://user-images.githubusercontent.com/17654553/103212591-80d76600-490b-11eb-9f96-7eae404e2075.png)

Please let me know if this makes sense or not, or if you would suggest a more pragmatic approach.

Sorry for the back and forth in the PR, Python is not my main language. Appreciated all the recommendations. 👍🏻

damiencarol reviewed (approved):
LGTM

valentijnscholten commented:
thanks for contributing and sticking with us through the recommendations :-)
